python/ray/autoscaler/local/node_provider.py

```diff
@@ -83,17 +96,64 @@ def put(self, worker_id, info):
                     f.write(json.dumps(workers))
 
 
+class OnPremCoordinatorState(ClusterState):
+    def __init__(self, lock_path, save_path, list_of_node_ips):
+        self.lock = RLock()
+        self.file_lock = FileLock(lock_path)
+        self.save_path = save_path
+
+        with self.lock:
+            with self.file_lock:
+                if os.path.exists(self.save_path):
+                    nodes = json.loads(open(self.save_path).read())
+                else:
+                    nodes = {}
+                logger.info(
+                    "OnPremCoordinatorState: "
+                    "Loaded on prem coordinator state: {}".format(nodes))
+
+                # Filter removed node ips.
+                for node_ip in list(nodes):
+                    if node_ip not in list_of_node_ips:
+                        del nodes[node_ip]
+
+                for node_ip in list_of_node_ips:
+                    if node_ip not in nodes:
+                        nodes[node_ip] = {
+                            "tags": {},
+                            "state": "terminated",
+                        }
+                assert len(nodes) == len(list_of_node_ips)
+                with open(self.save_path, "w") as f:
+                    logger.info(
+                        "OnPremCoordinatorState: "
+                        "Writing on prem coordinator state: {}".format(nodes))
+                    f.write(json.dumps(nodes))
+
+
 class LocalNodeProvider(NodeProvider):
     """NodeProvider for private/local clusters.
 
     `node_id` is overloaded to also be `node_ip` in this class.
+    When `cluster_name` is None, it coordinates multiple clusters.
```

Comment:
Can you expand on this? It could use a few paragraphs of description on how this works instead of a sentence.